### PR TITLE
Use Oracles for sparse select and prepare

### PR DESCRIPTION
**Pull request by fdmalone**
Use SelectOracle and PrepareOracle in sparse chemistry algorithm.

### CODE REVIEW

**Comment by fdmalone** [comment]
xref #546

**Comment by mpharrigan** [comment]
Can you add a description to the original post in the PR. This gets added to the commit history

File: qualtran/bloqs/chemistry/sparse/prepare.py [hunk 1]
```diff
@@ -77,39 +80,84 @@ class PrepareSparse(Bloq):
...
-                Register("pqrs", (self.num_spin_orb // 2 - 1).bit_length(), shape=(4,)),
-                Register("theta", 1),
-                Register("alpha", 1),
-                Register("beta", 1),
-                Register("swap_pq", 1),
-                Register("swap_rs", 1),
-                Register("swap_pqrs", 1),
-                Register("flag_1b", 1),
-            ]
+    def selection_registers(self) -> Tuple[SelectionRegister, ...]:
```
Comment (mpharrigan) [inline qualtran/bloqs/chemistry/sparse/prepare.py:86]: `Tuple[Register, ...]` is being returned here. Is this ok? If yes: update base class type annotation

File: qualtran/bloqs/chemistry/sparse/prepare.py [hunk 2]
```diff
@@ -77,39 +80,84 @@ class PrepareSparse(Bloq):
...
+                block_size - 1
             )  # A14
         qrom_cost = (Toffoli(), num_toff_qrom)
         if self.adjoint:
             return {
                 (PrepareUniformSuperposition(self.num_non_zero, self.num_bits_rot_aa), 1),
                 qrom_cost,
             }
-        swap_cost_state_prep = (CSwapApprox(num_bits_spat), 4 + 4)  # 2. pg 39
+        swap_cost_state_prep = (CSwap(num_bits_spat), 4 + 4)  # 2. pg 39
```
Comment (mpharrigan) [inline qualtran/bloqs/chemistry/sparse/prepare.py:160]: is this correct?

File: qualtran/bloqs/chemistry/sparse/prepare.py [hunk 3]
```diff
@@ -77,39 +80,84 @@ class PrepareSparse(Bloq):
...
-                Register("theta", 1),
-                Register("alpha", 1),
-                Register("beta", 1),
-                Register("swap_pq", 1),
-                Register("swap_rs", 1),
-                Register("swap_pqrs", 1),
-                Register("flag_1b", 1),
-            ]
+    def selection_registers(self) -> Tuple[SelectionRegister, ...]:
+        # issue here in that pqrs should not be reflected on.
```
Comment (mpharrigan) [inline qualtran/bloqs/chemistry/sparse/prepare.py:87]: link to github issue. Is this blocking the PR?

**Review by mpharrigan — COMMENTED** [review]
is this blocked by the referenced issue?

File: qualtran/bloqs/chemistry/sparse/prepare.py (hunk 2 above)
Comment (fdmalone) [reply qualtran/bloqs/chemistry/sparse/prepare.py:160]: The Approx -> regular swap? Yes (at least the approx cswap isn't referenced ever in the chemistry algorithms) so I think this is more appropriate.

File: qualtran/bloqs/chemistry/sparse/prepare.py (hunk 3 above)
Comment (fdmalone) [reply qualtran/bloqs/chemistry/sparse/prepare.py:87]: No it's not blocking, can revisit in future. I'll add the link

File: qualtran/bloqs/chemistry/sparse/prepare.py (hunk 1 above)
Comment (fdmalone) [reply qualtran/bloqs/chemistry/sparse/prepare.py:86]: Fixed.

File: qualtran/bloqs/chemistry/sparse/prepare.py (hunk 3 above)
Comment (fdmalone) [reply qualtran/bloqs/chemistry/sparse/prepare.py:87]: Done.

**Review by mpharrigan — APPROVED** [review]
lgtm

**Pull request merged**